### PR TITLE
Added the skeleton, data structures, sell function, and wake all up.

### PR DESCRIPTION
**Pull request by sagarghuge**
The code works with generating customer queues, creating threads,
waking all up.

### CODE REVIEW

**Review by rohansumant — COMMENTED** [review]
The basic skeleton seems good. We have most of the parts we need to get the final thing going. however, I have 2 suggestions:

1. The mutex should be a part of ticketSell struct as that is the struct where we are going to see contention between threads.

2. I am not sure what serve_t is but looking at line 42, it looks like that it is the clock. I think that clock should either be incremented in the simulator thread (convenient) or should have its own separate thread. Incrementing it in the seller thread seems incorrect.

Let me know what you guys think about this and if I have misunderstood anything.

File: ticketSelling.c [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+typedef struct customerQ {
+	customer *cust;
+	int front, rear;
+} customerQ;
+
+
+typedef struct ticketSell {
+	int next_avail;
+} ticketSell;
```
Comment (cortiz1) [inline ticketSelling.c:25]: suggestion to rename this to 'row'

File: ticketSelling.c [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+pthread_cond_t cond = PTHREAD_COND_INITIALIZER;
+pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static int serve_t = 0;
+// seller thread to serve one time slice (1 minute)
+void * sell(char *seller_type)
+{
+
+	while (serve_t < 60)
```
Comment (cortiz1) [inline ticketSelling.c:42]: this while loop should be in the simulator

File: ticketSelling.c [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+
+typedef struct ticketSell {
+	int next_avail;
+} ticketSell;
+
+customerQ *custQ;
+ticketSell tickets[10];
+
+float H_serve_time[] = {0.1, .2};
```
Comment (cortiz1) [inline ticketSelling.c:30]: why are these floats? i thought we were dealing with integers as minutes.

File: ticketSelling.c [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+	while (serve_t < 60)
+	{
+		pthread_mutex_lock(&mutex);
+		//pthread_cond_wait(&cond, &mutex);
+		
+		if(strcmp(seller_type, "H") == 0){
+			float serve_time  = H_serve_time[rand() % 2];
+			printf("Sleeping thread %s for time %f\n", seller_type, serve_time);
+			sleep(serve_time);
```
Comment (cortiz1) [inline ticketSelling.c:50]: I think this is confusing, from what i see you are sleeping here to simulate the serve time, but we actually put other sellers to sleep when they wait on a condition variable. i think that processing the sale should not actually sleep the thread but simply increment the simulation's clock by one.

File: ticketSelling.c [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static int serve_t = 0;
+// seller thread to serve one time slice (1 minute)
+void * sell(char *seller_type)
+{
+
+	while (serve_t < 60)
+	{
+		pthread_mutex_lock(&mutex);
```
Comment (cortiz1) [inline ticketSelling.c:44]: i think this is wrong, getting the lock here means we are totally serializing the sell function.

**Review by cortiz1 — COMMENTED** [review]
I agree with Rohan ticketSell should contain lock, also i suggest to rename this to 'row' since it represents one row.
also doesnt seem right to have the while loop inside the sell function as it implies that each thread has and uses their own clock.
I think we should have a 'clock' counter that is incremented by the simulator thread and the sell function checks that to determine time of serving.
Also using 'sleep' in the sell function seems to be adding a real time component that we dont need. also the sell function is entirely serialized by basically making it a big critical section.

i think the sell function needs a refactor, everything else looks good.

**Pull request merged**